src/compiler/checker.ts

```diff
@@ -6631,6 +6631,9 @@ namespace ts {
             else if (type.flags & TypeFlags.Intersection) {
                 return getIntersectionType(map((<IntersectionType>type).types, t => getTypeWithThisArgument(t, thisArgument, needApparentType)));
             }
+            else if (type.flags & TypeFlags.TypeParameter && thisArgument && (type as TypeParameter).isThisType) {
+                return getTypeWithThisArgument(createTypeReference(globalThisType, [type]), needApparentType ? instantiateType(thisArgument, createTypeMapper([type], [getApparentType(type)])) : thisArgument, needApparentType);

```

Comment:
1. can you format the conditional one more than one line? It's not readable as-is. (I would extract the middle argument to a temp.)
2. Why create a reference to `globalThisType`?